Add "Along the river" scenario

hessenfarmer opened the pull request.

hessenfarmer committed:
Added Along the river scenario

Noordfrees reviewed (commented):
I can't playtest this until next week, but already some comments here and inline:

- Please rename the map directory to `Along_The_River.wmf` (no whitespaces in directory names). 
- Do we want to keep the scenario *The Green Plateau*? If yes, please move it to `SP_Scenarios/` as well; otherwise delete it (in which case I'll add it to the Formerly Official Maps add-on).
- Can you also please add the new directory name `SP_Scenarios` to the exceptions in `src/editor/ui_menus/main_menu_load_map.cc` and `src/wui/mapdata.cc` (like for `MP_Scenarios`).

Noordfrees commented:
Oh and please update `.tx/config` (just copy the setting for The Green Plateau for the new scenario)

hessenfarmer commented:
@Nordfrees thanks for the quick review. 
Some of your remarks were already on my list (the reason why this is a draft), but I wanted to make this playtestable at the earliest stage, to get feedback. 
Will try to incorporate all changes tonight or the coming days.    

I think we could afford keeping the green plateau as well to make the plural in the folder name sensible ;-)

hessenfarmer committed:
cleaned up directories

hessenfarmer committed:
Added Translation Catch for new Singleplayer scenarios folder

hessenfarmer committed:
fix of elemental information

hessenfarmer committed:
added explaining texts to the objectives

hessenfarmer committed:
renamed pics to be more telling

hessenfarmer committed:
changed textdomain name

hessenfarmer committed:
cleaned up helper functions

hessenfarmer committed:
added new scenario to translation config

hessenfarmer committed:
made the messages of transport_wares number dependent

Commit:
4 files were automatically formatted.

hessenfarmer committed:
fix lua formatting

hessenfarmer committed:
Merge branch 'Sp_Scenario_River' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
made one fortress connectable

hessenfarmer committed:
ensured the pirate sentry will not vanish before we reach it

hessenfarmer commented:
Ok I implemented all review comments and added some code to ensure the relevant pirate sentry will exist when we see it.    
I thought of forbidding the opponents to attack each other. Opinions?

klaus-halfmann commented:
Checked out and will playtest this. Uhm, how can I start this scenario?

hessenfarmer committed:
corrected values in a comment

hessenfarmer commented:
> 
> 
> Checked out and will playtest this. Uhm, how can I start this scenario?

singleplayer --> new game --> singleplayer sceanrios --> Along the River

hessenfarmer committed:
Merge branch 'master' into Sp_Scenario_River

hessenfarmer committed:
Merge branch 'Sp_Scenario_River' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
another silly lookover

hessenfarmer committed:
added an objective to avoid having no objective for a long time.

klaus-halfmann commented:
I playtestet this for a while, but in my current status, two objectives are not "fullfilled" even though I think I did what was neede? pleate take a look [Along1.wgf.zip](https://github.com/widelands/widelands/files/7190201/Along1.wgf.zip)

hessenfarmer committed:
fixed wrong image

Noordfrees reviewed (commented):
I didn't get around to testing this yet, but a number of comments on the strings inline (sometimes several changes in one comment). Also please use Title Case consistently for objective and message headers, and the  `’` character for apostrophes.

klaus-halfmann commented:
Maybe you fixed this already?

[06:27:35.364 game] ERROR: [../src/scripting/lua_errors.cc:22] [string "scripting/infrastructure.lua"]:138: attempt to call a nil value (method 'set_wares')
[06:27:35.364 game] ERROR: Send message to all players and pause game

hessenfarmer committed:
apostroph fixes

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
textual fix

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
suggestion implemented

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
suggestion implemented

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
suggestion implemented

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
suggestion implemented

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
suggestion implemented

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
suggestion implemented

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
suggestion implemented

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
suggestion implemented

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer commented:
> 
> 
> Maybe you fixed this already?
> 
> [06:27:35.364 game] ERROR: [../src/scripting/lua_errors.cc:22] [string "scripting/infrastructure.lua"]:138: attempt to call a nil value (method 'set_wares')
> [06:27:35.364 game] ERROR: Send message to all players and pause game

yes that was fixed here.
https://github.com/widelands/widelands/pull/5062/commits/51bb84d993e9b97e66ed8155fa4873db16bd8565

hessenfarmer commented:
> 
> 
> I didn't get around to testing this yet, but a number of comments on the strings inline (sometimes several changes in one comment). Also please use Title Case consistently for objective and message headers, and the `’` character for apostrophes.

what do you mean by Title Case

hessenfarmer committed:
fixed further comments

hessenfarmer commented:
> 
> 
> I playtestet this for a while, but in my current status, two objectives are not "fullfilled" even though I think I did what was neede? pleate take a look [Along1.wgf.zip](https://github.com/widelands/widelands/files/7190201/Along1.wgf.zip)

the warehouse is in the wrong place. Will think about a more fuzzy logic to check for it 
for the right place se
[Along1a.zip](https://github.com/widelands/widelands/files/7190731/Along1a.zip)
e

klaus-halfmann commented:
OK, I "fixed" the warehous Issue. Nice idea with the showels. I now got stuck at the yellow player: Two citadells with only heros while runnimg out of iron, that was hard. I will retart again (now that I know where I can find the Iron).

Noordfrees commented:
> what do you mean by Title Case

The First Letter of Every Word Should Be Uppercase in a Title or Heading

klaus-halfmann commented:
Got this shortly after startup:
[00:00:02.563 real] INFO: **** END SOUND REPORT ****
[00:00:49.884 real] WARNING: New-style view packet not found. There may be strange effects regarding unseen areas.

hessenfarmer committed:
simplified fuction suggestion

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
syntax fix

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
simplified sysntax suggestion

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
simplified sysntax suggestion

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
textual fix

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
simplified syntax

hessenfarmer committed:
not working try to spawn soldiers additionally

hessenfarmer committed:
corrected function

klaus-halfmann commented:
Just found this crash, when batteing with the yellow player:
[06:31:54.805 game] ERROR: [../src/scripting/lua_errors.cc:22] [string "scripting/messages.lua"]:156: attempt to index a nil value (local 'message')
[06:31:54.805 game] ERROR: Send message to all players and pause game
...
[06:40:00.112 game] WARNING: Cmd_EnemyFlagAction::execute: wrong player target not seen or not hostile.
Assertion failed: (nr_icons_), function draw_overlay, file ../src/wui/inputqueuedisplay.cc, line 584.
Abort trap: 6
Will try to reproduce, laster

hessenfarmer commented:
> 
> 
> Just found this crash, when batteing with the yellow player:
> [06:31:54.805 game] ERROR: [../src/scripting/lua_errors.cc:22] [string "scripting/messages.lua"]:156: attempt to index a nil value (local 'message')
> [06:31:54.805 game] ERROR: Send message to all players and pause game
> ...
> [06:40:00.112 game] WARNING: Cmd_EnemyFlagAction::execute: wrong player target not seen or not hostile.
> Assertion failed: (nr_icons_), function draw_overlay, file ../src/wui/inputqueuedisplay.cc, line 584.
> Abort trap: 6
> Will try to reproduce, laster

which commit are you playing? do you have a savegame?

edit: found the issue and fixed it  see https://github.com/widelands/widelands/pull/5062/commits/935e60b29192095aec8434e3e48298034b5e303b

hessenfarmer committed:
text_fix

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
textual fix

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
Fixed Title Case

hessenfarmer commented:
> 
> 
> OK, I "fixed" the warehous Issue. Nice idea with the showels. I now got stuck at the yellow player: Two citadells with only heros while runnimg out of iron, that was hard. I will retart again (now that I know where I can find the Iron).

allowing multiple places for the warehouse is complicated, due to needing to know where the warehouse is exactly located for the shovels objective. I am thinking about placing the constructionsite for the player to ensure it is located in the correct place. Any opinions?

hessenfarmer committed:
fixed a search replace leftover

hessenfarmer committed:
fixed another legacy code issue

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer commented:
ok current version is not crashing anymore. took me 11:50 to beat this very nice but nor easy scenario

hessenfarmer committed:
Merge branch 'master' into Sp_Scenario_River

hessenfarmer committed:
Merge branch 'Sp_Scenario_River' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
forbade AI players fighting each other

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

klaus-halfmann commented:
I was finally able to play this to some end, but I never manager to get to seafaring. 
* How did I miss this?
* Shall I start again
Here some late savegames
[Along.zip](https://github.com/widelands/widelands/files/7273444/Along.zip)

hessenfarmer commented:
> 
> 
> I was finally able to play this to some end, but I never manager to get to seafaring.
> 
>     * How did I miss this?

the game you attached was created with a very old commit ( https://github.com/widelands/widelands/pull/5062/commits/bbaa255210e4590121d809d89bc1883f568f7b69) 

which means it had a lua coroutine error stopping the scripting (which you reported above) . So the seafarnig objective was not triggered. 
Unfortunately the scripting is saved with the map and the solution is indeed to start anew game.

> 
>     * Shall I start again
>       Here some late savegames
>       [Along.zip](https://github.com/widelands/widelands/files/7273444/Along.zip)

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

klaus-halfmann commented:
Playtesting looks better now, I "freed" the scout etc.
But now I hit:
[07:02:37.773 game] ERROR: [.../sp_scenario_river/src/scripting/lua_errors.cc:22] [string "scripting/functions.lua"]:25: attempt to index a nil value (global 'immovable')
[07:02:37.773 game] ERROR: Send message to all players and pause game

Do you need a savegame or such?

klaus-halfmann commented:
OPS,
Now I got an Assertion:
Assertion failed: (nr_icons_), function draw_overlay, file /Users/klaus/develop/widelands/sp_scenario_river/src/wui/inputqueuedisplay.cc, line 584

[Crash.txt.zip](https://github.com/widelands/widelands/files/7315827/Crash.txt.zip)
.

Noordfrees commented:
> OPS, Now I got an Assertion: Assertion failed: (nr_icons_), function draw_overlay, file /Users/klaus/develop/widelands/sp_scenario_river/src/wui/inputqueuedisplay.cc, line 584

That's #5052. I'm trying to investigate it but I still didn't manage to reproduce the bug

hessenfarmer commented:
> 
> 
> Playtesting looks better now, I "freed" the scout etc. But now I hit: [07:02:37.773 game] ERROR: [.../sp_scenario_river/src/scripting/lua_errors.cc:22] [string "scripting/functions.lua"]:25: attempt to index a nil value (global 'immovable') [07:02:37.773 game] ERROR: Send message to all players and pause game
> 
> Do you need a savegame or such?

savegame shortly before the lua error would indeed help a lot

klaus-halfmann commented:
> > Playtesting looks better now, I "freed" the scout etc. But now I hit: [07:02:37.773 game] ERROR: [.../sp_scenario_river/src/scripting/lua_errors.cc:22] [string "scripting/functions.lua"]:25: attempt to index a nil value (global 'immovable') [07:02:37.773 game] ERROR: Send message to all players and pause game
> > Do you need a savegame or such?
> 
> savegame shortly before the lua error would indeed help a lot

Here you are (this is acutally when it happed)
[AlongLuaError.wgf.zip](https://github.com/widelands/widelands/files/7317008/AlongLuaError.wgf.zip)

hessenfarmer committed:
fixed a bug

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
fixed textdomain pushing

klaus-halfmann commented:
I think the basic mechanis of this are ok and I could approv this, but I think some improvements are missig:
* The correct / wrong placement of the warehouse should be commentet like
 "You build the warehouse, lets check the sourrounding how we may proceed"
 "This warehouse is not at the optimal location please go more to the west/east/north/south"

Fo the last two enemies ther should be a hint like:
* "You will need to collect at least 15 (20, 25) completly trained heros at the border, so please be patient"

hessenfarmer committed:
Merge branch 'master' into Sp_Scenario_River

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
made warehouse objective near the pirates brook more flexible.

Commit:
'data/maps/SP_Scenarios/Along_The_River.wmf/scripting/init.lua' was automatically formatted.

hessenfarmer commented:
ok I now allowed the warehouse to be placed in a region of radius 8 west of the originally foreseen place. 
It is properly detected and used for the ongoing script. So I would consider this issue fixed.  
About the hint with the heroes, I think a player will learn this once he tried it too early. No need to provide a walkthrough right in the script. If he manages to get that far I believe we could consider him a good player anyway and do not need to bore him with such hints.

hessenfarmer committed:
typo fix

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
grammar correction

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
adressed review

hessenfarmer committed:
Merge branch 'Sp_Scenario_River' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

hessenfarmer committed:
consistency textual change

klaus-halfmann commented:
@hessenfarmer would some more testplays help you?

Noordfrees commented:
Code LGTM now :) 

I didn't yet have time to play this through though :( If @klaus-halfmann could play the scenario through from start to end with the latest commit and finds no bugs I'd say let's have it :)

hessenfarmer commented:
> @hessenfarmer would some more testplays help you?

Actually I tested the latest change with the warehouse in the debug console, whether it works. Only other functional change was the change in the arrays of the swamps and the for loops instead of the while loops. 
so playtesting could be limited to play until the swamps got removed and the objective to defeat the pirates is displayed. 
Nevertheless i wouldn't object if anybody seeks the fun to play to the end ;-)

Noordfrees commented:
Well, I never trust scenario authors to test their own scenarios :wink: Bugs are more reliably found by people doing things the author never expected, and any change that's been tested only from console can also introduce typos or whatever :)

hessenfarmer commented:
> 
> 
> Well, I never trust scenario authors to test their own scenarios 😉 Bugs are more reliably found by people doing things the author never expected, and any change that's been tested only from console can also introduce typos or whatever :)

valid point ;-) so a full playtest would be needed by @klaus-halfmann . Have fun!

klaus-halfmann commented:
Still got that ware house wrong @hessenfarmer 
[Warehouse.wgf.zip](https://github.com/widelands/widelands/files/7520980/Warehouse.wgf.zip)
please check

hessenfarmer commented:
> 
> 
> Still got that ware house wrong @hessenfarmer [Warehouse.wgf.zip](https://github.com/widelands/widelands/files/7520980/Warehouse.wgf.zip) please check

According to the data the objective to build a warehouse is fulfilled. so it should be removed from the objectives list. 
for new objectives you need to defeat the red player as well.

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into Sp_Scenario_River

klaus-halfmann reviewed (approved):
I played this now until only the yellow player was left, but It took me to long to beat the other too. So
When the yellow play went Into Attack mode I has no Chance to win against it.

But yes, this can go in now. 
@hessenfarmer will you remove the add-on with the same now (or kind of deprecate it)?

hessenfarmer commented:
> 
> 
> I played this now until only the yellow player was left, but It took me to long to beat the other too. So When the yellow play went Into Attack mode I has no Chance to win against it.
> 
> But yes, this can go in now. @hessenfarmer will you remove the add-on with the same now (or kind of deprecate it)?

Thanks for the review and hours of playtesting. 
Regarding the Addon this was meant to be for the great community assumably using still Widelands 1.0. So I will update it with the latest changes after this is merged to give the community the same version as in current development,  and I will delete the Addon as soon as 1.1 is out.

hessenfarmer committed:
Merge branch 'master' into Sp_Scenario_River

hessenfarmer commented:
FYI Addon is already updated.
